hc5/src/main/java/feign/hc5/AsyncApacheHttp5Client.java

```diff
@@ -0,0 +1,183 @@
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package feign.hc5;
+
+import org.apache.hc.client5.http.async.methods.SimpleHttpRequest;
+import org.apache.hc.client5.http.async.methods.SimpleHttpResponse;
+import org.apache.hc.client5.http.config.Configurable;
+import org.apache.hc.client5.http.config.RequestConfig;
+import org.apache.hc.client5.http.impl.async.CloseableHttpAsyncClient;
+import org.apache.hc.client5.http.impl.async.HttpAsyncClients;
+import org.apache.hc.client5.http.protocol.HttpClientContext;
+import org.apache.hc.core5.concurrent.FutureCallback;
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.Header;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import feign.*;
+import feign.Request.Options;
+
+/**
+ * This module directs Feign's http requests to Apache's
+ * <a href="https://hc.apache.org/httpcomponents-client-5.0.x/index.html">HttpClient 5</a>. Ex.
+ *
+ * <pre>
+ * GitHub github = Feign.builder().client(new ApacheHttp5Client()).target(GitHub.class,
+ * "https://api.github.com");
+ */
+/*
+ */
+public final class AsyncApacheHttp5Client<C> implements AsyncClient<C> {
+
+  private static final String ACCEPT_HEADER_NAME = "Accept";
+
+  private final CloseableHttpAsyncClient client;
+
+  public AsyncApacheHttp5Client() {
+    this(HttpAsyncClients.custom().build());
+  }
+
+  public AsyncApacheHttp5Client(CloseableHttpAsyncClient client) {
+    this.client = client;
+    client.start();
+  }
+
+  @Override
+  public CompletableFuture<Response> execute(Request request,
+                                             Options options,
+                                             Optional<C> requestContext) {
+    final SimpleHttpRequest httpUriRequest = toClassicHttpRequest(request, options);
+
+    final CompletableFuture<Response> result = new CompletableFuture<>();
+    final FutureCallback<SimpleHttpResponse> callback = new FutureCallback<SimpleHttpResponse>() {
+
+      @Override
+      public void completed(SimpleHttpResponse httpResponse) {
+        result.complete(toFeignResponse(httpResponse, request));
+      }
+
+      @Override
+      public void failed(Exception ex) {
+        result.completeExceptionally(ex);
+      }
+
+      @Override
+      public void cancelled() {
+        result.cancel(false);
+      }
+    };
+
+    client.execute(httpUriRequest, callback);
+
+    return result;
+  }
+
+  protected HttpClientContext configureTimeouts(Request.Options options) {
+    final HttpClientContext context = new HttpClientContext();
```

Comment:
```suggestion
    final HttpClientContext context = requestContext.orElseGet(HttpClientContext::new)
```
Use the client supplied context, else make a new default just for this request.